File: src/Wollax.Cupel/ContextItem.cs
using System.Text.Json.Serialization;

namespace Wollax.Cupel;

/// <summary>
/// Immutable data model representing a single piece of context in the pipeline.
/// Every scorer, slicer, and placer operates on ContextItem instances.
/// </summary>
public sealed record ContextItem
{
    /// <summary>
    /// The full textual content of this context item. Must not be null.
    /// Used directly by deduplication (content-based) and by scorers that inspect text.
    /// </summary>
    [JsonPropertyName("content")]
    public required string Content { get; init; }

    /// <summary>
    /// Token count for this item, measured in the target model's tokenization unit.
    /// Must be zero or positive — items with negative tokens are excluded by the pipeline
    /// before scoring. Zero-token items are always included when budget permits.
    /// </summary>
    [JsonPropertyName("tokens")]
    public required int Tokens { get; init; }

    /// <summary>
    /// Semantic category of this context item. Controls kind-based scoring weights and
    /// quota allocation. Defaults to <see cref="ContextKind.Message"/>.
    /// </summary>
    [JsonPropertyName("kind")]
    public ContextKind Kind { get; init; } = ContextKind.Message;

    /// <summary>
    /// Origin source of this context item — e.g. chat history, retrieval, or tool output.
    /// Used by source-aware scorers. Defaults to <see cref="ContextSource.Chat"/>.
    /// </summary>
    [JsonPropertyName("source")]
    public ContextSource Source { get; init; } = ContextSource.Chat;

    /// <summary>
    /// Optional explicit priority for this item (higher value = higher importance).
    /// When set, the <see cref="Scoring.PriorityScorer"/> uses this value directly.
    /// When null, the PriorityScorer treats the item as having no declared priority.
    /// </summary>
    [JsonPropertyName("priority")]
    public int? Priority { get; init; }

    /// <summary>
    /// Descriptive tags attached to this item. Used by the <see cref="Scoring.TagScorer"/>
    /// to apply per-tag weight boosts. Empty list is valid and means no tag scoring applies.
    /// </summary>
    [JsonPropertyName("tags")]
    public IReadOnlyList<string> Tags { get; init; } = [];

    /// <summary>
    /// Arbitrary key-value metadata for application-defined use.
    /// Not used by any built-in scorer or slicer. Safe to leave empty.
    /// </summary>
    [JsonPropertyName("metadata")]
    public IReadOnlyDictionary<string, object?> Metadata { get; init; }
        = new Dictionary<string, object?>();

    /// <summary>
    /// Wall-clock time when this item was created, observed, or last updated.
    /// Used by the <see cref="Scoring.RecencyScorer"/> to rank recent items higher.
    /// When null, the RecencyScorer treats this item as having no temporal signal.
    /// </summary>
    [JsonPropertyName("timestamp")]
    public DateTimeOffset? Timestamp { get; init; }

    /// <summary>
    /// Self-reported hint for future-relevance scoring, conventionally in the range [0.0, 1.0].
    /// Higher values indicate the item is more likely to be relevant in future turns.
    /// Used by the <see cref="Scoring.ReflexiveScorer"/>. When null, no reflexive score is applied.
    /// </summary>
    [JsonPropertyName("futureRelevanceHint")]
    public double? FutureRelevanceHint { get; init; }

    /// <summary>
    /// When true, this item is always included in the output regardless of its score or budget.
    /// Pinned items are never passed through the slicer — they are merged after slicing.
    /// Pinned items override quota caps if their combined tokens exceed a quota limit.
    /// </summary>
    [JsonPropertyName("pinned")]
    public bool Pinned { get; init; }

    /// <summary>
    /// Token count of this item before any summarization or truncation was applied.
    /// When set, callers can compare <see cref="Tokens"/> to <c>OriginalTokens</c> to detect
    /// lossy compression. When null, no compression has been recorded. Must be zero or positive.
    /// </summary>
    [JsonPropertyName("originalTokens")]
    public int? OriginalTokens { get; init; }
}

File: crates/cupel/src/model/context_item.rs
use std::collections::HashMap;

use chrono::{DateTime, Utc};

use crate::CupelError;
use crate::model::{ContextKind, ContextSource};

/// An immutable record representing a single piece of context in the pipeline.
///
/// Constructed via [`ContextItemBuilder`]. All fields are private with public accessor methods.
#[derive(Debug, Clone, PartialEq)]
pub struct ContextItem {
    content: String,
    tokens: i64,
    kind: ContextKind,
    source: ContextSource,
    priority: Option<i64>,
    tags: Vec<String>,
    metadata: HashMap<String, String>,
    timestamp: Option<DateTime<Utc>>,
    future_relevance_hint: Option<f64>,
    pinned: bool,
    original_tokens: Option<i64>,
}

impl ContextItem {
    /// The textual content of this context item.
    pub fn content(&self) -> &str {
        &self.content
    }

    /// The token count for this context item.
    pub fn tokens(&self) -> i64 {
        self.tokens
    }

    /// The kind of context item.
    pub fn kind(&self) -> &ContextKind {
        &self.kind
    }

    /// The origin of this context item.
    pub fn source(&self) -> &ContextSource {
        &self.source
    }

    /// Optional priority override. Higher values indicate greater importance.
    pub fn priority(&self) -> Option<i64> {
        self.priority
    }

    /// Descriptive tags for filtering and scoring.
    pub fn tags(&self) -> &[String] {
        &self.tags
    }

    /// Arbitrary key-value metadata, opaque to the pipeline.
    pub fn metadata(&self) -> &HashMap<String, String> {
        &self.metadata
    }

    /// When this context item was created or observed.
    pub fn timestamp(&self) -> Option<DateTime<Utc>> {
        self.timestamp
    }

    /// Hint for future relevance scoring, conventionally in [0.0, 1.0].
    pub fn future_relevance_hint(&self) -> Option<f64> {
        self.future_relevance_hint
    }

    /// Whether this item is pinned (bypasses scoring and slicing).
    pub fn pinned(&self) -> bool {
        self.pinned
    }

    /// The original token count before any external summarization or truncation.
    pub fn original_tokens(&self) -> Option<i64> {
        self.original_tokens
    }
}

/// Builder for constructing [`ContextItem`] instances.
#[derive(Debug)]
pub struct ContextItemBuilder {
    content: String,
    tokens: i64,
    kind: ContextKind,
    source: ContextSource,
    priority: Option<i64>,
    tags: Vec<String>,
    metadata: HashMap<String, String>,
    timestamp: Option<DateTime<Utc>>,
    future_relevance_hint: Option<f64>,
    pinned: bool,
    original_tokens: Option<i64>,
}

impl ContextItemBuilder {
    /// Creates a new builder with the required `content` and `tokens` fields.
    pub fn new(content: impl Into<String>, tokens: i64) -> Self {
        Self {
            content: content.into(),
            tokens,
            kind: ContextKind::default(),
            source: ContextSource::default(),
            priority: None,
            tags: Vec::new(),
            metadata: HashMap::new(),
            timestamp: None,
            future_relevance_hint: None,
            pinned: false,
            original_tokens: None,
        }
    }

    /// Sets the kind of context item.
    pub fn kind(mut self, kind: ContextKind) -> Self {
        self.kind = kind;
        self
    }

    /// Sets the source of this context item.
    pub fn source(mut self, source: ContextSource) -> Self {
        self.source = source;
        self
    }

    /// Sets the priority override.
    pub fn priority(mut self, priority: i64) -> Self {
        self.priority = Some(priority);
        self
    }

    /// Sets the descriptive tags.
    pub fn tags(mut self, tags: Vec<String>) -> Self {
        self.tags = tags;
        self
    }

    /// Sets the metadata map.
    pub fn metadata(mut self, metadata: HashMap<String, String>) -> Self {
        self.metadata = metadata;
        self
    }

    /// Sets the timestamp.
    pub fn timestamp(mut self, timestamp: DateTime<Utc>) -> Self {
        self.timestamp = Some(timestamp);
        self
    }

    /// Sets the future relevance hint.
    pub fn future_relevance_hint(mut self, hint: f64) -> Self {
        self.future_relevance_hint = Some(hint);
        self
    }

    /// Sets whether this item is pinned.
    pub fn pinned(mut self, pinned: bool) -> Self {
        self.pinned = pinned;
        self
    }

    /// Sets the original token count.
    pub fn original_tokens(mut self, tokens: i64) -> Self {
        self.original_tokens = Some(tokens);
        self
    }

    /// Builds the [`ContextItem`], validating that content is non-empty.
    pub fn build(self) -> Result<ContextItem, CupelError> {
        if self.content.is_empty() {
            return Err(CupelError::EmptyContent);
        }

        Ok(ContextItem {
            content: self.content,
            tokens: self.tokens,
            kind: self.kind,
            source: self.source,
            priority: self.priority,
            tags: self.tags,
            metadata: self.metadata,
            timestamp: self.timestamp,
            future_relevance_hint: self.future_relevance_hint,
            pinned: self.pinned,
            original_tokens: self.original_tokens,
        })
    }
}
